bugsnag-plugin-android-ndk/src/main/jni/utils/threads.c

```diff
@@ -0,0 +1,199 @@
+ * https://github.com/torvalds/linux/blob/v5.13/fs/proc/array.c#L132-L143
+ */
+static const char *const task_state_array[] = {
+    "R running", "S sleeping", "D disk sleep", "T stopped", "t tracing stop",
+    "X dead",    "Z zombie",   "P parked",     "I idle",
+};
+
+void get_task_state_description(const char code, char *dest) {
+  for (size_t index = 0; index < (sizeof(task_state_array) / sizeof(char *));
+       index++) {
+    if (task_state_array[index][0] == code) {
+      bsg_strcpy(dest, &task_state_array[index][2]);
+      return;
+    }
+  }
+
+  // if we reached here we failed to map the code, store the single-character
+  // code as a fallback
+  dest[0] = code;
+  dest[1] = '\0';
+}
+
+/**
+ * Parses the content of a /proc/self/task/{tid}/stat file, as documented in:
+ * https://man7.org/linux/man-pages/man5/proc.5.html
+ *
+ * We are only interested in the first three fields which are the:
+ * 1) TID   (numeric)
+ * 2) Name  (in parenthesis)
+ * 3) State (single character)
+ *
+ * Unfortunately thread names can contain spaces, making tools like `strtok`
+ * unsuitable for this process. As a result we parse using a simple single pass
+ * state-based loop.
+ *
+ * *WARNING* `content` is modified by this method under normal operation
+ *
+ * @return true if the thread-data was parsed, false if not
+ */
+bool parse_stat_content(bsg_thread *dest, char *content, size_t len) {
+  enum stat_parse_state {
+    PARSE_ID,
+    PARSE_NAME_START,
+    PARSE_NAME_CONTENT,
+    PARSE_NAME_END,
+    PARSE_STATUS,
+    PARSE_DONE
+  };
+
+  enum stat_parse_state state = PARSE_ID;
+
+  size_t i = 0;
+  size_t name_length = 0;
+  while (i < len) {
+    char current = content[i];
+
+    switch (state) {
+    case PARSE_ID:
+      if (current == ' ') {
+        // we create a terminator for the numeric parse
+        content[i] = '\0';
+        // atoi is async-safe, strtol is not guaranteed to be
+        dest->id = atoi(content);
+        state = PARSE_NAME_START;
+      }
+      break;
+    case PARSE_NAME_START:
+      if (current == '(') {
+        state = PARSE_NAME_CONTENT;
+      }
+      break;
+    case PARSE_NAME_CONTENT:
+      if (current == ')') {
+        state = PARSE_NAME_END;
+      } else if (name_length < sizeof(dest->name)) {
+        dest->name[name_length] = current;
+        name_length++;
+      }
+      break;
+    case PARSE_NAME_END:
+      dest->name[name_length] = '\0';
+      if (current == ' ') {
+        state = PARSE_STATUS;
+      }
+      break;
+    case PARSE_STATUS:
+      get_task_state_description(current, dest->state);
+      state = PARSE_DONE;
+      break;
+    case PARSE_DONE:
+      goto end;
+    }
+
+    i += 1;
+  }
+
+end:
+  // success if we hit the DONE marker
+  return state == PARSE_DONE;
+}
+
+/*
+ * Reads the /stat file fields we are looking for (TID, Name, State) into `dest`
+ * with a signal-safe approach.
+ */
+bool read_thread_state(bsg_thread *dest, const char *tid) {
```

Comment:
The helper functions in this file have fairly generic names, which could cause a conflict somewhere. They should be marked static so they aren't visible elsewhere.
```suggestion
static bool read_thread_state(bsg_thread *dest, const char *tid) {
```